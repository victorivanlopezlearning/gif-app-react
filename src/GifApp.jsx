import { useState } from 'react';
import Form from './components/Form';
import ErrorMessage from './components/ErrorMessage';

const GifApp = () => {

  const [categories, setCategories] = useState([]);
  const [errorMesagge, setErrorMesagge] = useState('');

  const onAddCategory = (newCategory) => {

    const categoryExist = categories.find((category) => category.toUpperCase() === newCategory.toUpperCase());
    if (categoryExist) {
      setErrorMesagge(`${newCategory} ya est√° listado. Por favor, indica uno diferente.`);
      return;
    }
    setErrorMesagge('');
    setCategories([newCategory, ...categories]);
  };

  return (
    <>
      <h1>Buscador de GIFs</h1>

      <Form
        onAddCategory={onAddCategory}
      />
      {errorMesagge && <ErrorMessage errorMesagge={errorMesagge} />}

      <ol>
        {categories?.map(category => (
          <li key={category}>{category}</li>
        ))}
      </ol>
    </>
  )
}

export default GifApp;