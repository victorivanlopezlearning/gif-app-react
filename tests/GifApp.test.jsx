import { fireEvent, render, screen } from '@testing-library/react';
import GifApp from '../src/GifApp';

describe('Tests in <GifApp.jsx />', () => {

  const category = 'Metal Gear Solid';

  test('Should match with the snapshot', () => {
    const { container } = render(<GifApp />);
    expect(container).toMatchSnapshot();
  });

  test('Should show heading H1 with the expected text', () => {
    render(<GifApp />);

    const { textContent } = screen.getByRole('heading', { level: 1 });

    expect(textContent).toBe('Buscador de GIFs');
  });

  test('Should not show error message and Heading h2 of the category title', () => {

    render(<GifApp />);

    const input = screen.getByRole('textbox');
    const form = screen.getByRole('form');

    fireEvent.input(input, { target: { value: category } });
    fireEvent.submit(form);

    const { textContent } = screen.getByRole('heading', { level: 2 });

    expect(screen).not.toBe(`${category} ya está listado. Por favor, indica uno diferente.`);
    expect(textContent).toBe(category);
  });

  test('Should show error message when repeating category', () => {

    render(<GifApp />);

    const input = screen.getByRole('textbox');
    const form = screen.getByRole('form');

    fireEvent.input(input, { target: { value: category } });
    fireEvent.submit(form);
    fireEvent.input(input, { target: { value: category } });
    fireEvent.submit(form);

    expect(screen.getByText(`${category} ya está listado. Por favor, indica uno diferente.`)).toBeTruthy();
  });
});