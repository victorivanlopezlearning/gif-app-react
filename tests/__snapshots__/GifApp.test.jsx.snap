// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <GifApp.jsx /> Should match with the snapshot 1`] = `
<div>
  <h1>
    Buscador de GIFs
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar GIFs"
      type="text"
      value=""
    />
  </form>
</div>
`;
